docs/src/design/stack/crypto_ops.md

```diff
@@ -0,0 +1,122 @@
+
+The effect of this operation on the rest of the stack is:
+* **No change** starting from position $0$.
+
+## MRUPDATE
+The `MRUPDATE` operation computes a new root of a Merkle tree where a node at the specified position is updated to the specified value.
+    
+The stack is expected to be arranged as follows (from the top):
+- depth of the node, 1 element
+- index of the node, 1 element
+- old value of the node, 4 element
+- new value of the node, 4 element
+- current root of the tree, 4 elements
+
+The Merkle path for the node is expected to be provided by the prover non-deterministically (via advice sets). At the end of the operation, the old node value is replaced with the old root value computed based on the provided path, the new node value is replaced by the new root value computed based on the same path. Everything else on the stack remains the same. The diagram below illustrates this graphically.
+
+![mrupdate](../../assets/design/stack/crypto_ops/MRUPDATE.png)
+
+In the above, $r$ (located in the helper register $h_0$) is the row address from the hash chiplet set by the prover nondeterministically.
+
+For the `MRUPDATE` operation, we define input and output values as follows:
+
+$$
+v_{inputold} = \alpha_0 + \alpha_1 \cdot op_{mruold} + \alpha_2 \cdot r + \alpha_3 \cdot i + \sum_{j=0}^3\alpha_{j + 8} \cdot s_{5 - j}
+$$
+
+$$
+v_{outputold} = \alpha_0 + \alpha_1 \cdot op_{rethash} + \alpha_2 \cdot (r + 8 \cdot d - 1) + \sum_{j=0}^3\alpha_{j + 8} \cdot s_{13 - j}
+$$
+
+$$
+v_{inputnew} = \alpha_0 + \alpha_1 \cdot op_{mrunew} + \alpha_2 \cdot (r + 8 \cdot d) + \alpha_3 \cdot i + \sum_{j=0}^3\alpha_{j + 8} \cdot s_{9 - j}
+$$
+
+$$
+v_{outputnew} = \alpha_0 + \alpha_1 \cdot op_{rethash} + \alpha_2 \cdot (r + 2 \cdot 8 \cdot d - 1) + \sum_{j=0}^3\alpha_{j + 8} \cdot s_{9 - j}'
+$$
+
+In the above, the first two expressions correspond to inputs and outputs for verifying the Merkle path between the old node value and the old tree root, while the last two expressions correspond to inputs and outputs for verifying the Merkle path between the new node value and the new tree root. Hash chiplet ensures the same set of sibling nodes are uses in both of these computations.
+
+The $op_{mruold}$, $op_{mrunew}$, and $op_{rethash}$ are the unique [operation label](../chiplets/main.md#operation-labels) used by the above computations.
+
+> $$
+b_{chip}' \cdot v_{inputold} \cdot v_{outputold} \cdot v_{inputnew} \cdot v_{outputnew} = b_{chip} \text{ | degree } = 5
+$$
+
+The above constrain enforces that the specified input and output rows for both, the old and the new node/root combinations, must be present in the trace of the hash chiplet, and that they must be exactly $8 \cdot d - 1$ rows apart, where $d$ is the depth of the node. It also ensure that the computation for the old node/root combination is immediately followed by the computation for the new node/root combination.
```

Comment:
nit: "The above constraint enforces... It also ensures..."

Reply:
Fixed!